### PR TITLE
machine: add USBDevice.Attach and USBDevice.Detach

### PR DESCRIPTION
**Pull request by sago35**
This PR adds `machine.USBDev.Attach()` and `machine.USBDev.Detach()`, exposing
USB soft-connect control (the D+ pull-up) to user code.

**Motivation**

The USB device is attached to the bus automatically during startup, before
user code has had a chance to finish its USB configuration (`usb.VendorID` /
`usb.ProductID` / `usb.Product`, additional interfaces registered through
`machine.ConfigureUSBEndpoint`, ...). For composite devices such as keyboards
this means the host may enumerate a half-configured device: identifiers and
interfaces set from `main()` can arrive after the host has already read the
descriptors.

With this API, an application or library can stay detached until its
configuration is complete, so the host enumerates the finished device exactly
once:

```go
func init() {
      machine.USBDev.Detach() // stay invisible to the host for now
}

func main() {
      usb.Product = "my-keyboard"
      // ... register HID interfaces, set VID/PID, ...
      machine.USBDev.Attach() // now let the host enumerate
}
```

It can also be used to force re-enumeration without replugging the cable
(Detach() → Attach()).

Prior art

Soft-connect control is a well-established primitive in embedded USB stacks,
exposed for exactly these use cases (delaying enumeration until configuration
is complete, and forcing re-enumeration from software):

- Arduino: USBDevice.detach() / USBDevice.attach() (AVR and SAMD
cores).
- TinyUSB: tud_disconnect() / tud_connect().
- Zephyr: usb_dc_detach() / usb_dc_attach() in the device-controller
driver API.
- STM32 HAL: HAL_PCD_DevDisconnect() / HAL_PCD_DevConnect().

The Attach/Detach naming follows the USB specification's bus-state wording
(a device is "attached" to the bus), which is also where the SAMD register
bit name (CTRLB.DETACH) comes from — and matches the Arduino API, easing
migration for users coming from that ecosystem.

Implementation

One-register operation per chip family, following each reference manual:

```
┌────────────────────┬─────────────────────────────┐
│        Chip        │          Mechanism          │
├────────────────────┼─────────────────────────────┤
│ atsamd21, atsamd51 │ USB_DEVICE.CTRLB DETACH bit │
├────────────────────┼─────────────────────────────┤
│ nrf52840           │ USBD.USBPULLUP              │
├────────────────────┼─────────────────────────────┤
│ rp2040, rp2350     │ SIE_CTRL PULLUP_EN bit      │
└────────────────────┴─────────────────────────────┘
```

The ESP32 variants are not covered: their USB is a Serial/JTAG peripheral
with no equivalent soft-connect concept. Existing behavior is unchanged —
the methods are purely additive.

Testing

- Compile-tested a program calling both methods for all five chip families
(xiao-ble, xiao, wioterminal, pico, pico2).
- Verified on hardware with a XIAO nRF52840 Sense running a Vial keyboard
firmware (tinygo-keyboard)
that detaches in a package init() and attaches after configuration:
the host enumerates the device once, with the complete configuration
(CDC + keyboard/mouse/raw HID interfaces and a custom ProductID).**

### CODE REVIEW

File: src/machine/machine_rp2040_usb.go [hunk 1]
```diff
@@ -46,6 +46,21 @@ func (dev *USBDevice) Configure(config UARTConfig) {
 	rp.USBCTRL_REGS.SIE_CTRL.SetBits(rp.USBCTRL_REGS_SIE_CTRL_PULLUP_EN)
 }
 
+// Attach connects the device to the USB bus by enabling the DP pull-up,
+// allowing the host to detect and enumerate it. It can be used together with
+// Detach to delay enumeration until the USB configuration (device
+// identifiers, classes, ...) is complete.
+func (dev *USBDevice) Attach() {
+	rp.USBCTRL_REGS.SIE_CTRL.SetBits(rp.USBCTRL_REGS_SIE_CTRL_PULLUP_EN)
```
Comment (rdon-key) [inline src/machine/machine_rp2040_usb.go:54]: RP2040/RP2350 provide atomic bitmask set/clear aliases at +0x2000 and
+0x3000 (RP2040 datasheet 2.1.2, p.18:
https://datasheets.raspberrypi.com/rp2040/rp2040-datasheet.pdf), which
modify only the selected bits. SIE_CTRL is also touched from the ISR,
so this avoids the read-modify-write. Could we use those aliases here
for PULLUP_EN?

**Review by rdon-key — COMMENTED** [review]
I left one comment about using the RP2040/RP2350 atomic set/clear aliases for SIE_CTRL.PULLUP_EN.

File: src/machine/machine_rp2040_usb.go (hunk 1 above)
Comment (sago35) [reply src/machine/machine_rp2040_usb.go:54]: Thanks for the great suggestion. I think we can definitely use it here, but it would probably be better handled in a separate PR—for example, by integrating it into implementations like `SetBits()` / `ClearBits()`/ ....

However, this suggested approach might only work for registers and not for standard RAM. If that’s the case, we might need to consider a slightly more complex approach.

Since that change would have a broad impact, it's likely better to target the `dev` branch right after the upcoming release, rather than squeezing it into `TinyGo 0.42`.

Would you be open to creating a separate PR for that?

Comment (sago35) [reply src/machine/machine_rp2040_usb.go:54]: I created Issue #5564

Comment (rdon-key) [reply src/machine/machine_rp2040_usb.go:54]: Understood. I'll look into a more general solution, although it may take some time.

For this PR, could we keep the change local and use the atomic aliases directly?

 For example:

```go
// USBCTRL_REGS aliases for atomic bitmask set/clear (datasheet section 2.1.2).
var (
	usbSet = (*rp.USBCTRL_REGS_Type)(unsafe.Pointer(
		uintptr(unsafe.Pointer(rp.USBCTRL_REGS)) + 0x2000))
	usbClr = (*rp.USBCTRL_REGS_Type)(unsafe.Pointer(
		uintptr(unsafe.Pointer(rp.USBCTRL_REGS)) + 0x3000))
)

func (dev *USBDevice) Attach() {
	usbSet.SIE_CTRL.Set(rp.USBCTRL_REGS_SIE_CTRL_PULLUP_EN)
}

func (dev *USBDevice) Detach() {
	usbClr.SIE_CTRL.Set(rp.USBCTRL_REGS_SIE_CTRL_PULLUP_EN)
}
```

And similarly for RP2350.

Comment (sago35) [reply src/machine/machine_rp2040_usb.go:54]: I don't think there is a need to change to a new method in this PR.

This code is a new change and has no impact on other parts. Furthermore, by using `SetBits()` and similar functions, the internal implementation will be updated to a better form. There is no reason to leave open the possibility of an implementation that differs from the future `SetBits()`.

Comment (rdon-key) [reply src/machine/machine_rp2040_usb.go:54]: Thanks. Understood. I still think this should use the atomic write even temporarily, but it's your call.

**Comment by deadprogram** [comment]
Hello @sago35 thanks for working on this.

Here are a couple of items extracted from an automated code review.

1. nRF52840: Detach() isn't sticky. The IRQ handler unconditionally re-enables the pull-up on every USBEVENT/EVENTCAUSE_READY:

```
// machine_nrf52840_usb.go:105-107
initEndpoint(0, usb.ENDPOINT_TYPE_CONTROL)
nrf.USBD.USBPULLUP.Set(1)
```

Unlike the other four chips, the pull-up here is owned by the ISR rather than Configure(), so a later power-ready event can silently re-attach. Could a detached bool on USBDevice guard that line? (The Detach-in-init → Attach-in-main flow you tested wouldn't necessarily hit it.)

2. Missing platforms. usb.go also builds for stm32f4/f7/h7, and esp32c3/c6/xx declare their own USBDevice, so machine.USBDev.Attach() won't compile there. stm32 isn't mentioned in the description but is the easy win — dctlSDIS (machine_stm32_otgfs_usb.go:98) and DCTL_SDIS (machine_stm32h7_usb.go:120) already exist and are already toggled in Configure. For esp32, machine_esp32xx_usb.go:65-68 sets the no-op-stub precedent (SetStallEPIn etc.), so no-op methods there would keep user code portable.

What do you think?

**Comment by sago35** [comment]
I will check later.

**Comment by sago35** [comment]
@deadprogram 
Both points are addressed, split into three commits so they can be judged (or dropped) individually. The branch is also rebased onto the current dev.

1. nRF52840: sticky Detach (c16ff7b1)

Added a `usbDetached` flag. Detach() sets it, Attach() clears it, and the IRQ handler now only re-enables the DP pull-up on a power-ready event when the flag is not set, so a later USBEVENT/EVENTCAUSE_READY can no longer silently re-attach the device. The flag is written before the register in both methods, so an event racing with Detach() still ends up detached.

2. Missing platforms (858a56fa)

- stm32f4 / stm32f7 / stm32h7: implemented Attach/Detach with the DCTL soft-disconnect bit (dctlSDIS / DCTL_SDIS). Note that I have no stm32 hardware at hand, so these are compile-tested only — but they reuse the exact soft-disconnect sequence Configure already performs on these chips, so they should behave the same.
- esp32c3 / esp32c6 / esp32s3: added no-op stubs next to the existing SetStallEPIn etc. stubs, so machine.USBDev.Attach() stays portable. The fixed-function USB Serial/JTAG controller has no software-controlled soft-connect.

3. Refactor (e23373ef, optional)

Replaced the direct soft-connect register writes inside Configure with the equivalent dev.Attach() / dev.Detach() calls on atsamd21, atsamd51, rp2040, rp2350 and the three stm32 families. This one is purely cosmetic — feel free to ask me to drop it if you prefer the explicit register writes in Configure.

One deliberate exception: the `USBPULLUP.Set(0)` at the start of the nRF52840 Configure is not replaced with Detach(), because that would set the sticky flag and prevent the IRQ handler from ever attaching during normal enumeration.

Verified by compiling a program that calls `machine.USBDev.Detach()` / `Attach()` for all affected targets: circuitplay-express, feather-m4, feather-nrf52840, pico, pico2, feather-stm32f405, nucleo-f722ze, nucleo-h753zi, esp32c3-12f, esp32s3-generic and xiao-esp32c6.

**Comment by deadprogram** [comment]
Thank you very much for adding this @sago35 now squash/merging.
